File: app/src/main/java/com/example/myapplication/MainActivity.java
package com.example.myapplication;

import android.Manifest;
import android.annotation.SuppressLint;

import androidx.annotation.NonNull;
import kotlinupdatepackage.KotlinDeleteFunction;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.StrictMode;

import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.ProtocolException;
import java.net.URL;
import java.util.ArrayList;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;


public class MainActivity extends AppCompatActivity {

    public static TextView message;
    public static ProgressBar pb;
    public String[] tcodes;
    public static final String EXTRA_MESSAGE = "com.example.MyApplication.MESSAGE";
    private EditText username;
    private EditText passwor;
    private TextView current_version;
    private Context mContext;
    public final String APP_TAG = "MyCustomApp";
    private static final int DELAY = 10000;
    public boolean inactive = true;
    public static boolean inactive2 = true;

    Handler mHideHandler = new Handler();
    Runnable mHideRunnable = new Runnable() {
        @Override
        public void run() {
            if(inactive2) {
                finishAffinity();
            } else {
                mHideHandler.postDelayed(mHideRunnable, DELAY);
                inactive2 = true;
            }
        }
    };

    public static void setInactivityFromAnotherActivity(boolean newStatus){
        inactive2 = newStatus;
    }

    @Override
    public void onUserInteraction() {
        super.onUserInteraction();
        inactive2 = false;
    }



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        this.mContext = this;
        username = (EditText) findViewById(R.id.badgenum);
        username.requestFocus();

        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
        StrictMode.setThreadPolicy(policy);
        Button btn = (Button) findViewById(R.id.button_login);
        Button update_btn = (Button) findViewById(R.id.updatebutton);
        pb = (ProgressBar)findViewById(R.id.progressBar);
        pb.setVisibility(View.INVISIBLE);
        //mHideHandler.postDelayed(mHideRunnable, DELAY);

        new checkForUpdateAsyncTask().execute();

        update_btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                KotlinDeleteFunction.Companion.update_method1(MainActivity.this);
            }
        });

        btn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View view) {


                new YourAsyncTask().execute();

                //sendMessage(view);

            }
        });


    }


    private Uri fileUri;
    @SuppressLint("NewApi")
    private void deleteDownloadedFiles() throws FileNotFoundException {

        requestForPermission();

            getFile();


        return;

    }

    private void requestForPermission() {
        ActivityCompat.requestPermissions(
                this,
                new String[]{
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE
                },
                101
        );
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private void getFile() {
        mContext = MainActivity.this;
        File file;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            file = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString());
        } else {
            file = new File(Environment.getExternalStorageDirectory().toString() + "/Download/");
        }
        File[] allfiles = null;
        allfiles = file.listFiles();
        if (file.exists()) {
            for (File f:allfiles) {
                if (f.getAbsoluteFile().toString().contains("app-debug")){


                }
            }
        }
        if (allfiles != null) {
            Toast.makeText(this, "length is " + allfiles.length, Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, "Array is null", Toast.LENGTH_SHORT).show();
        }
    }



    private class checkForUpdateAsyncTask extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... args) {

            KotlinDeleteFunction.Companion.update_method1(MainActivity.this);
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            pb.setVisibility(View.INVISIBLE);
            super.onPostExecute(result);
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            pb.setVisibility(View.VISIBLE);
        }
    }

    private class YourAsyncTask extends AsyncTask<Void, Void, Void> {

        AlertDialog.Builder alertDialog;
        public String build_utl(String usr,String pass){
            pb.setVisibility(View.VISIBLE);
            String init = "http://172.16.206.19/REST_API/Home/Barcode_Login?user_name=" + usr + "&pwd=" + pass;
            return init;
        }

        @NonNull
        public ArrayList<ArrayList<String>> get_login_result_from_url(String url) {


            ArrayList<ArrayList<String>> functions_Description = new ArrayList<ArrayList<String>>();


            try {
                URL Url = new URL(url);
                HttpURLConnection conn = (HttpURLConnection)Url.openConnection();
                conn.setRequestMethod("GET");
                conn.setRequestProperty("Accept", "application/json");
                if (conn.getResponseCode() != 200) {
                    throw new RuntimeException("Failed : HTTP Error code : "
                            + conn.getResponseCode());
                }
                InputStreamReader in = new InputStreamReader(conn.getInputStream());
                BufferedReader br = new BufferedReader(in);
                String line = br.readLine();
                JSONArray jsonarray = new JSONArray(line);
                for (int i=0; i < jsonarray.length(); i++)
                {
                    try {
                        JSONObject oneObject = jsonarray.getJSONObject(i);
                        // Pulling items from the array
                        String oneObjectsItem = oneObject.getString("Function");
                        String TWOObjectsItem = oneObject.getString("Description");
                        ArrayList<String> functions = new ArrayList<>();
                        functions.add(oneObjectsItem);
                        functions.add(TWOObjectsItem);
                        functions_Description.add(functions);


                    } catch (JSONException e) {
                        JSONObject oneObject = jsonarray.getJSONObject(i);
                        String oneObjectsItem = oneObject.getString("Function");
                        ArrayList<String> functions = new ArrayList<>();
                        functions.add(oneObjectsItem);
                        functions.add("");
                        functions_Description.add(functions);
                    }
                }
            } catch (MalformedURLException e) {

                e.printStackTrace();
                delegate.processFinish(e.getMessage().toString());
            } catch (ProtocolException e) {
                delegate.processFinish(e.getMessage().toString());
                e.printStackTrace();
            } catch (IOException e) {
                delegate.processFinish(e.getMessage().toString());
                e.printStackTrace();
            } catch (JSONException e) {
                delegate.processFinish(e.getMessage().toString());
                e.printStackTrace();
            }


            return functions_Description;
        }

        public AsyncResponse delegate = null;

        String error_check = null;
        @Override
        protected Void doInBackground(Void... args) {
            pb = (ProgressBar)findViewById(R.id.progressBar);
            // Do something in response to button
            TextView bn = (TextView)findViewById(R.id.badgenum);
            TextView bnp = (TextView)findViewById(R.id.editTextTextPassword);
            String name = bn.getText().toString();
            int k = 0;
            String url = build_utl(bn.getText().toString(),bnp.getText().toString());
            ArrayList<ArrayList<String>> functions = get_login_result_from_url(url);

            Intent intent = new Intent(MainActivity.this, Main_Screen.class);
            intent.putExtra(EXTRA_MESSAGE, name);
            intent.putExtra("key",functions);
            String messg = functions.get(0).toString();
            if(!functions.get(0).toString().contains("User")){
                startActivity(intent);


            }
            else{
                error_check= "Invalid Username/Password";

            }
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            pb.setVisibility(View.INVISIBLE);
            super.onPostExecute(result);
            if(error_check != null){
                alertDialog.setTitle("Error");
                alertDialog.setMessage(error_check);
                alertDialog.setPositiveButton("ok",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.dismiss();
                            }
                        });
                AlertDialog orderError = alertDialog.create();
                orderError.show();

                error_check = null;
            }

        }

        @Override
        protected void onPreExecute() {
                alertDialog = new AlertDialog.Builder(MainActivity.this);
            super.onPreExecute();
            pb.setVisibility(View.VISIBLE);
        }



    }

    public interface AsyncResponse {
        void processFinish(String output);
    }



}

File: app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.widget.Toast
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import java.io.IOException
import java.net.URL

class MainActivity : AppCompatActivity() {
    private lateinit var username: EditText
    private lateinit var mContext: MainActivity
    private lateinit var UsernameInput: EditText
    private lateinit var PasswordInput: EditText
    private lateinit var loginbtn: Button
    private lateinit var updateBtn: Button
    private lateinit var badgeCaps: String
    private lateinit var pb: ProgressBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.constraintlayout)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        mContext = this
        username = findViewById<View>(R.id.badgenum) as EditText
        username.requestFocus()
        UsernameInput = findViewById(R.id.badgenum)
        PasswordInput = findViewById(R.id.edMainPass)
        loginbtn = findViewById(R.id.button_login)
        updateBtn = findViewById(R.id.updatebutton)
        pb = findViewById(R.id.progressBar)

        updateBtn.setOnClickListener {
            runBlocking {
                GlobalScope.launch {
                    val url: URL = URL("http://172.16.206.19/REST_API/debug/output-metadata.json")
                    val outputMetadata = url.readText()
                    val OMObject = JSONTokener(outputMetadata).nextValue() as JSONObject
                    var apkVersion = OMObject.getJSONArray("elements")
                    var apkcurrentVersion = apkVersion.getJSONObject(0).getString("versionCode")

                    val apkversionName = apkVersion.getJSONObject(0).getString("versionName")
                    val currentVersion = getCurrentVersion(applicationContext)

                    if (apkcurrentVersion != currentVersion) {

                    }


                }
            }
        }

        loginbtn.setOnClickListener {
            if (isNetworkAvailable(mContext)) {
                if(username.text.isNullOrEmpty() || PasswordInput.text.isNullOrEmpty()){
                    Toast.makeText(mContext, "InpuT Password or username", Toast.LENGTH_LONG).show()
                }
                else{
                    badgeCaps = username.text.toString().uppercase()
                    execute_get_login_result(badgeCaps, PasswordInput.text.toString(), pb)
                }


            }
            else{
                Toast.makeText(mContext, "No network connection available!", Toast.LENGTH_LONG).show()
            }

        }

    }



    fun isNetworkAvailable(context: Context): Boolean {
        val connectivityManager =
            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = connectivityManager.activeNetwork ?: return false
        val networkCapabilities =
            connectivityManager.getNetworkCapabilities(network) ?: return false
        return networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }

    fun getCurrentVersion(context: Context): String? {
        try {
            val packageInfo: PackageInfo =
                context.packageManager.getPackageInfo(context.packageName, 0)
            return packageInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            return "N/A"
        }
    }

    private fun get_login_result(userName: String?, password: String?): String? {
        var result = String()
        if (userName.isNullOrEmpty() || password.isNullOrEmpty()) {
            throw IllegalArgumentException("Please enter username AND password")
        }
        val loginUrl =
            "http://172.16.206.19/REST_API/Home/Barcode_Login?user_name=${userName}&pwd=${password}"


        return try {
            val url = URL(loginUrl)
            url.readText()  // Fetch the response as a String
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    private fun setProgressBar(v: View, show: Boolean) {
        if (show) {
            v.visibility = View.VISIBLE
            loginbtn.isEnabled = false
        } else {
            v.visibility = View.GONE
            loginbtn.isEnabled = true
        }
    }

    private fun execute_get_login_result(userName: String?, password: String?, progressBar: View) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                // Show the progress bar and disable the button
                setProgressBar(progressBar, true)

                // Perform the network request in the IO dispatcher
                val result = withContext(Dispatchers.IO) {
                    get_login_result(userName, password)
                }
                setProgressBar(progressBar, true)
                // Handle the result
                result?.let { responseString ->

                    val function =
                        JSONObject(JSONArray(responseString).getString(0)).getString("Function")
                    if (function != "User not Registered") {

                        val intent = Intent(mContext, Main_Screen::class.java)
                        intent.putExtra("Badge", badgeCaps)
                        intent.putExtra("Functions", responseString)
                        startActivity(intent)
                    } else {
                        val builder = AlertDialog.Builder(mContext)
                        builder.setTitle("Attention")
                        builder.setMessage(function)
                        builder.setPositiveButton("Ok") { dialog, which ->
                            UsernameInput.text.clear()
                            PasswordInput.text.clear()
                            UsernameInput.requestFocus()
                        }
                        builder.show()
                    }
                }
            }            catch(e:Exception) {}
            finally {
                setProgressBar(pb, false)
            }

        }
    }
}